core/src/main/java/com/google/cloud/sql/core/LazyRefreshStrategy.java

```diff
@@ -0,0 +1,119 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import java.time.Instant;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** RefreshStrategy that implements the lazy refresh strategy. */
+public class LazyRefreshStrategy implements RefreshStrategy {
+  private final Logger logger = LoggerFactory.getLogger(LazyRefreshStrategy.class);
+
+  private final String name;
+  private final Supplier<ConnectionInfo> refreshOperation;
+
+  private final Object connectionInfoGuard = new Object();
+
+  @GuardedBy("connectionInfoGuard")
+  private ConnectionInfo connectionInfo;
+
+  @GuardedBy("connectionInfoGuard")
+  private boolean closed;
+
+  /** Creates a new LazyRefreshStrategy instance. */
+  public LazyRefreshStrategy(String name, Supplier<ConnectionInfo> refreshOperation) {
+    this.name = name;
+    this.refreshOperation = refreshOperation;
+  }
+
+  @Override
+  public ConnectionInfo getConnectionInfo(long timeoutMs) {
+    synchronized (connectionInfoGuard) {
+      if (closed) {
+        throw new IllegalStateException("Named connection closed");
+      }
+
+      if (connectionInfo == null) {
+        logger.debug(
+            String.format(
+                "[%s] Client certificate has expired. Starting next refresh operation immediately.",
+                name));
+        fetchConnectionInfo();
+      }
+      if (Instant.now().isAfter(connectionInfo.getExpiration())) {
+        logger.debug(
+            String.format(
+                "[%s] Client certificate has expired. Starting next refresh operation immediately.",
+                name));
+        fetchConnectionInfo();
+      }
+      return connectionInfo;
+    }
+  }
+
+  private void fetchConnectionInfo() {
+    synchronized (connectionInfoGuard) {
+      logger.debug(String.format("[%s] Refresh Operation: Starting refresh operation.", name));
+      try {
+        this.connectionInfo = this.refreshOperation.get();
+      } catch (TerminalException e) {
+        logger.debug(String.format("[%s] Refresh Operation: Failed! No retry.", name), e);
+        throw e;
+      } catch (Exception e) {
+        throw new RuntimeException(String.format("[%s] Refresh Operation: Failed!", name), e);
+      }
+
+      logger.debug(
+          String.format(
+              "[%s] Refresh Operation: Completed refresh with new certificate expiration at %s.",
+              name, connectionInfo.getExpiration().toString()));
+    }
+  }
+
+  /** Force a new refresh of the instance data if the client certificate has expired. */
+  @Override
+  public void forceRefresh() {
+    // invalidate connectionInfo so that the next call to getConectionInfo() will retrieve
+    // fetch new data.
```

Comment:
nit: choose one of retrieve/fetch they are synonyms
```suggestion
    // invalidate connectionInfo so that the next call to getConectionInfo() will retrieve
    // fetch new data.
```

Reply:
Fixed.